Riverlea - add Stream entity

ufundo opened the pull request:
Overview
----------------------------------------
Move the Stream definitions in Riverlea from hard-coded to a new entity.

This enables a "Previewer" which allows you to switch streams locally, *without* changing the site settings.

Paves the way for future UI to a) pick a stream based on what it looks like on your site; b) editing and creating custom streams through the UI (AKA "Customiser").

Before
----------------------------------------
- available streams are hard-coded in `riverlea.php`

After
----------------------------------------
- streams are provided by a new entity RiverleaStream
- streams can be viewed with Search Kit, edited with FormBuilder etc
- external extensions can provide RiverLea streams using managed entity
- dark mode option can be set at the stream level (though will fallback to the global setting if unset)
- you can create a stream on one site and then export it using the ManagedEntity export (like you can with SearchKits)
- you can open the console and type `CRM.riverlea.previewer()` to open a previewer widget, which allows you to switch between streams on the fly
- streams can include `custom_css`, which allows for easy adding fixes. these are then cached via assetbuilder (essentially using the stream name and last modified as key) which overcomes the performance concerns with previous attempt at Customiser 
- streams can include `vars` defined using php, which could then be validated/updated automatically etc

https://github.com/user-attachments/assets/ca59d58d-1502-4b03-81ce-01d551728601


(note all previewer changes are on the frontend and local to your current session - the site theme setting isn't touched. It *does* require starting with a Riverlea theme selected)

Work in progress
----------------------------------------
Need to add upgrader to create the new table on existing installs. I think this goes in the core upgrade step for core extensions?

Would be good to get some feedback on how it's working / overall schema.

Known Limitation
----------------------------------------
- Previewer can't handle the `civicrm.css` override in Thames and Walbrook (it can only switch the var values on the fly). At this stage it's a dev tool, so I think it's reasonable. (I also think the overrides in Thames and Walbrook don't need to be in `civicrm.css` - they could be moved into the stream specific css file)
- Asset builder cache buster on `river.css` will only track changes to the entity, but if you use `css_file` and modify the file it won't necessarily know. Again, I think this is reasonable. It's not expected to edit css files directly in production - and you (will soon) be able to edit `custom_css` on the entity if you need to.

Technical Details
----------------------------------------
- Use a WebComponent `<civi-riverlea-previewer>`. I'm not sure how we feel about these but I think it's very nice for a self contained component that can use the API and talk to the browser.

ufundo commented:
(Just a note, if testing locally on a pre-existing site you will need to uninstall and then reinstall Riverlea in order to create the DB table)

vingle commented:
I'm not in a position (skills-wise) to comment on the code changes here - I can't comment on other consequences they might have. I can only really look at this from a product perspective:

Firstly a previewer is definitely really nice - ThemeTest would benefit from that (I spend quantifiable time moving back and forth from Display Setting). 

But this is more than that…

At present:
 - the path to creating custom streams is to create a subtheme extension, pointing to RiverLea. It's in the docs and it's a way to make Streams that survive upgrades. 
 - from a css designer's perspective that's an appealing approach as it's relatively straightforward (ie low-coding skills) to clone a subtheme extension, then find and replace the name to your Stream name. As a method it survives upgrades to RiverLea (unless you rename all RL's variables, lol). It can then be moved from dev to production site, stored on a git repo, version controlled, offered to the public, etc
 
 With this proposal:
  - a Stream is an entity… where are its files? What if it wants to bundle fonts or images? Would it still be a subtheme extension, but just that it would create a new entity on creation? Delete on uninstall?
  
And the benefit of all of this change is the previewer? Or Something else ("paves the way for future UI to a) pick a stream based on what it looks like on your site" - ie in a realtime preview?).

Side point: as RiverLea flows towards the big C of Civi Core - maybe it's name makes less and less sense? It's no longer a theme like Greenwich and more Civi's theme engine. So maybe - at least if it becomes an entity in core, its name should be something generic like CRMThemer, or at least more generic like 'river' (as river -> stream, is helpful naming to emphasise the cascade order of css).

ufundo commented:
Hey @vingle 

With this approach I see two ways to make new streams:

1) "No Code" => use the UI Customiser, hit an export button and get a Managed Entity text file. This can be
  a) copied and pasted into another site
  b) included in an extension 
It's the same format as used for sharing Search Kits, so relatively familiar to folks who've worked with that.

2) Code => similar to what you describe, you can basically copy the existing files for one of the existing streams into your own extension, change the `minetta` to `yourstreamname` and then you can work with the `.css` files directly. We could make a template extension to simplify the process. Previously you had to register the theme using `hook_civicrm_themes` - now you have to register it by creating a `MyTheme.mgd.php` but the contents are very similar. I can make template extension if you like?

The only advantage implemented in this PR is the previewer.

But it's a fairly small step from here to: 
- a stream-picker UI where you can choose which theme you want to use, with a preview before you activate it
- add snippets for hot-fixes
- set colours and other vars through the UI
- copy a stream through the UI, make some edits, flick back and forth between the original and your edited version

vingle commented:
> My thinking was that we don't need to create another hierarchy, and we should use stream for all the variations within River Lea. A stream = River Lea Theme = super easy to maintain theme.

Fair point.

> (I've started seeing it the other way round, the stream is the little drops of ink you add to the river at the beginning, and then the river takes that and spreads it all over the UI :))

Nooooh! 😅 as the ‘crm-c-text’ variable / drop flows from the River source, it first takes one colour from core, then as it branches into one Stream it takes a different colour, while in a second Stream no new variable/drop is defined so it stays the same (ie the order has to follow the cascade..)

> (without having to add css classes programmatically, which would be hard I think).

That makes sense

ufundo commented:
Further possibility with this kind of model: automating contrast checks between vars we know need to contrast. IE messages in the editor like "your crm-c-primary and crm-c-primary-text currently have contrast rating 5:1 which meets AA but you should increase for AAA"

vingle commented:
> automating contrast checks 

Sold!

ufundo commented:
I think we need thel change to `_riverlea_is_active` in this PR whether or not we go the entity route, so I've made a separate PR for it: https://github.com/civicrm/civicrm-core/pull/32263

ufundo committed:
riverlea - add Stream entity

ufundo committed:
riverlea - generate stream name from label

ufundo committed:
riverlea - clear theme cache after editing streams

ufundo committed:
riverlea - add managed records for base streams

ufundo commented:
I've narrowed this down to just the entity plumbing.

Listing, Previewing + Editing are separated to https://github.com/civicrm/civicrm-core/pull/32344

ufundo commented:
retest this please

colemanw commented:
@ufundo some test failures here...

ufundo commented:
tests should pass now @colemanw 

 ( a static/non-static function got dropped in rebasing somehow :grimacing: )

colemanw commented on ext/riverlea/CRM/riverlea/Upgrader.php:
```diff
@@ -0,0 +1,137 @@
+ */
+class CRM_riverlea_Upgrader extends \CRM_Extension_Upgrader_Base {
+
+  // By convention, functions that look like "function upgrade_NNNN()" are
```
@ufundo Looks like an upgrader is needed to add the new sql table (and all this boilerplate clutter can be deleted too).

I recently updated the docs about how to do so: https://docs.civicrm.org/dev/en/latest/step-by-step/create-entity/#5b-add-upgrade-step-for-extensions-used-in-production

ufundo replied:
ah thanks that's what I was after!

colemanw commented on ext/riverlea/schema/RiverleaStream.entityType.php:
```diff
@@ -0,0 +1,137 @@
+      'input_type' => 'Select',
+      'description' => ts('Extension that provides this stream.'),
+      'default' => NULL,
+      // @todo why not work?
```
Why not work?

ufundo replied:
I think because the `civicrm_extension` table is slightly odd.

ufundo replied:
I think it's kind of immaterial for now... for the time being this will always be provided by a `*.mgd.php` and it won't make sense to edit it.

colemanw replied:
@ufundo which part doesn't work? The FK index in SQL? Or the join in Api4. If the former, it's now possible to declare an EntityRef without the FK index like so:
https://github.com/civicrm/civicrm-core/blob/d1a8138f23cfda44503d8ac8077a1e756e949a12/schema/Core/UserJob.entityType.php#L158-L164

ufundo replied:
This still doesn't work in SearchKit / Formbuilder. I can't a) join Stream records to Extension records, b) in-place edit that field, c) add an afform filter.

I think maybe it's because `file` is not the primary key on `civicrm_extension` ? But we need to use a machine name (rather than `id`) because this is typically set in a managed record ( https://github.com/civicrm/civicrm-core/pull/32127/files#r2104616568 ).

I think for now it is fine to not have the entity_reference for now?

ufundo committed:
riverlea - add stream entity upgrader

ufundo commented on ext/riverlea/managed/RiverleaStream_HackneyBrook.mgd.php:
```diff
@@ -0,0 +1,28 @@
+        'label' => E::ts('Hackney Brook'),
+        'description' => 'named after the Hackney Brook, a tributary of the River Lea that ran through Finsbury Park',
+        'is_reserved' => TRUE,
+        'extension' => 'riverlea',
```
where `extension` is set

ufundo committed:
riverlea - update docs for creating a new stream

ufundo commented:
> Separately - could you update the ReadMe in this PR? 

Good thinking. I've had a go.

> Both to describe some of the new capabilities like the inspector previewer, and structural changes

This PR is just the pure plumbing, so it doesn't add any new features. But it does change how you add a new stream. I've updated the guidance and added a template example `*.mgd.php` file. It is a little bit clunky to do the setup by hand because there is no UI at this stage. With the UI , you can create the Stream through the UI first, and then export it (the same way it works for creating and distributing SearchKits)

ufundo commented:
Bump @colemanw

ufundo committed:
riverlea - update managed streams for clearer file paths

vingle commented on ext/riverlea/README.md:
```diff
@@ -103,53 +94,30 @@ For instance, to give all contribution page buttons rounded corners, you could a
-1. Inside the `/streams/` directory is an example stream called `empty`. Duplicate this and rename it the name of your stream.
-2. In riverlea.php add a theme array to the function `riverlea_civicrm_themes(&$themes)`.
-3. Edit `/streams/[streamname]/css/_variables.css` with your custom css variables. You can link to other CSS files, fonts or images in this file - inside the stream.
+Currently the best way to create a new stream is by adding a managed record to an extension.
```
There's maybe a missing step here like…

"If you're not adding this to your existing extension, first create one [following the Civi Docs](https://docs.civicrm.org/dev/en/latest/extensions/)? (plus to make clear it doesn't need to be a theme extension any more)."

ufundo replied:
Added

aydun commented:
> The architecture before was using Civi's subtheme functions - but I'm assuming this PR ends that? I was going to suggest mentioning sub-themes once at the start to help people, but it's true there's no loadable RiverLea theme so maybe it's better they use the new term.

On the topic of sub-thingies ... does this let us create a "sub-stream": a stream that inherits from another stream and applies overrides?  This is different to modifying a full clone to create a new stream.

> "No Code" => use the UI Customiser, hit an export button and get a Managed Entity text file. This can be

There could be 2 versions of this: 
1) the Managed Entity file is the entire definition - the original stream provided a useful starting and now I'm on my own.
2) the Managed Entity file specifies the name of the parent stream with only the variables that are different - I want all the improvements as the parent stream develops, but overlaid with my tweaks.

vingle commented:
> does this let us create a "sub-stream": a stream that inherits from another stream and applies overrides? 

That's what the goal is with the Customiser (still a draft), forking an existing core stream (a brook?!). I'm not sure if that's available in this PR tho. 

I agree with the principle should be to only store in a new stream/sub-stream the minimum that's changed. 

The customiser UI will also only expose a limited number of variables - one of the roadmap jobs with RL is how to simplify it to allow fewer variables to have more impact. Most people don't want to understand how five different background colours work with each other across Civi; that's RL implementing Greenwich+Shoreditch faithfully; most people expect to have one bg colour that cascades to other background regions with tints automatically.

ufundo commented:
Good questions.

> The architecture before was using Civi's subtheme functions - but I'm assuming this PR ends that?

It still uses the same `search_order` cascade, which is the sense that the streams could be called "sub themes" of `_riverlea_core_`. But a) AFAIK nothing outside of riverlea uses the term "sub-theme" for the cascade, that's a Riverlea-ism; b) it's under the hood, and not that relevant to user experience of picking a stream.

> On the topic of sub-thingies ... does this let us create a "sub-stream": a stream that inherits from another stream and applies overrides? This is different to modifying a full clone to create a new stream.
> There could be 2 versions of this:
> 
>     the Managed Entity file is the entire definition - the original stream provided a useful starting and now I'm on my own.
>     the Managed Entity file specifies the name of the parent stream with only the variables that are different - I want all the improvements as the parent stream develops, but overlaid with my tweaks.
> the principle should be to only store in a new stream/sub-stream the minimum that's changed.

So at the moment I have been working on the copy-and-edit approach. There's no direct support for this. I actually started with a `parent_id` to set streams as parents/children of other streams - but on working on the Customiser I got the sense this is actually not that helpful. I think it complicates more than it benefits, and that flat is better than nested here, because

- **inheritance leaves a tangled trail for stream editors:**
  - _"why are these icons in this one particular place "purple?! oh it's because the primary colour theme colour is set purple in parent theme, I've overridden that in my child theme to blue, but the parent theme had a slightly different shade of purple for primary-icons."_
  - _"oh I can really easily change some small thing about this stream.. arghhh! I didnt realise this would affect a child theme I made a long time ago for some very small specific purpose"_ 

- **inheritance makes life difficult for automatic upgraders.** part of the rationale of using a Stream entity is that it is possible to write upgraders if there are structural changes (e.g. to variable names). The names used for this variable changed in this release => no worries we can write an upgrader to update the definitions of existing streams. Upgraders are generally much safer and more straightforward if you can look at each record in isolation "if a stream has this combo of 3 variables, xyz needs to change". But when you have inheritance it's very hard to say if a stream (including all its parents) has this combo of 3 variables. You can't assume updating each stream can be done record by record, without worrying about the order.

- ** inheritance makes upstream development harder ** "I want all the improvements as the parent stream develops, but overlaid with my tweaks." would be ideal. But in practice it's hard to avoid clashes between my improvements upstream and your tweaks. If everyone starts making sub-themes of Walbrook, I think in reality it would quickly become too scary to make any changes to Walbrook itself, for fear of how it would cascade.

- ** The focus of Riverlea development going forward should be improving ALL streams ** My sense is that the next stages of Riverlea are improving the core cascades, with things that benefit *all* streams - the starter ones and the ones other folks make. There might be some little flourishes in Minetta etc... but you shouldn't miss out on much with the copy-and-amend model. 

- ** Streams should be SIMPLE ** - my sense is most streams should have a small set of variables (like, 20?) which then cascade automatically to provide a consistent experience throughout the UI. It may be you copy 18 of those variables from Thames to get the starting look, and then change 2 or 3. Maybe you set 20 directly. At any rate I think folks should be able to style CiviCRM WITHOUT having to get to grips with dozens and dozens of variables, and how these interact with dozens more in the parent stream, and the parent of the parent etc

- As a general rule, reconciling layers of multi-dimensional config is hard. "only store in a new stream the minimum that's changed" is a nice principle, but from that principle stems `git`.... We don't want to create git for civicrm themes...

Notwithstanding the above:
- Nothing about this change stops you from hooking into `hook_civicrm_managed` directly, as previously described in the README. (Though I think that doesn't actually work that well at the moment, because of the dynamic variable loading in order to respect dark mode settings.)
- It would still be very much possible to add inheritance between streams in the future. 


> most people expect to have one bg colour that cascades to other background regions with tints automatically

100% this.

ufundo committed:
riverlea - add note that it's early days for overridding vars; link to extension docs

ufundo commented:
One use case I can see for inheritance:

I have my "org theme" which has 20 vars set. But then I want to do a dozen "campaign themes" which use 19 vars but set the primary colour differently.

If I want to make a change to the button-roundness across the board, how?

I think it's relevantly different that I am the maintainer of the parent and child themes... they are a cluster. But it would be nice to avoid having to change that button-roundness var on every one of my dozen campaign themes.

One way this can be handled with the structure in this PR is:
- put the basic 20 vars in `my_ext/stream/main.css`
- set `css_file` / `file_prefix` to point at that file for all of my campaign themes
- set the campaign-specific vars in the stream `vars` directly

That way all streams use the same css file source, but can tweak at the entity level. You can update the css file once and it will affect all the campaign streams. 

In this sense there's a bit of inheritance built in within each stream, which can be leveraged for this sort of thing.

ufundo commented:
Further drawback of inheritance... you end up with versioned dependency issues.

I'm hoping that people will share nice streams they have made for others to use. This is much easier if the stream definition is self-contained -- NOT "my porpoise-themed stream is a sub-stream of Rich's dolphin stream, but you need to make sure you have `dolphin` version at least 1.2a but it doesn't work with v2".

ufundo commented:
> build front-end sites using Framework Z... so they make a stream that does 90% of the integration work which they ship with all their Civi

I think if its a frontend theme used outside of Civi, then that could integrate with Riverlea using vars etc WITHOUT being a stream (more along the lines of option 1). It's a more manual integration... but if you are a shop building something more bespoke then I think that's reasonable.  But I don't think streams themselves should be over-engineered for this kind of use case (and I think you could probably make it work in a way that allows you to overlay tweaks using streams - i.e. a combo of Option 1 and 3).

> I guess it would just mean you would treat a sub-stream as a core stream in the Customiser UI?

There's an `is_reserved` flag on streams. The UI doesn't let you edit those. That's the only difference between "core" streams and any others atm.
